File: algorithms/dijkstra.js
import { Graph } from '../data-sctructures/index.js';

export function dijkstra(graph, source) {
  const visited = [];

  const distance = {};
  const previous = {};

  for (const vertice of graph.vertices) {
    distance[vertice] = Infinity;
    previous[vertice] = null;
  }

  distance[source] = 0;

  let node = source;

  while (visited.length < graph.vertices.length) {
    // Get all neighbors, which wasn't already visited
    const unvisitedNeighbors = graph
      .getNeighbors(node)
      .filter(neighbor => !visited.includes(neighbor));

    // Get through all unvisited neighbors and check, if path from current node
    // is better, than their current
    for (const neighbor of unvisitedNeighbors) {
      const weight = graph.getWeight(node, neighbor);

      if (distance[node] + weight < distance[neighbor]) {
        distance[neighbor] = distance[node] + weight;
        previous[neighbor] = node;
      }
    }

    // Mark current node as visited
    visited.push(node);

    // Find unvisited node with smallest distance
    let minWeight = Infinity;
    let minNode = null;

    for (const key in distance) {
      if (visited.includes(key)) continue;

      if (distance[key] < minWeight) {
        minWeight = distance[key];
        minNode = key;
      }
    }

    node = minNode;
  }

  return { distance, previous };
}

const graph = new Graph(
  ['start', 'a', 'b', 'finish'],
  [
    { source: 'start', destination: 'a', weight: 6 },
    { source: 'start', destination: 'b', weight: 2 },
    { source: 'a', destination: 'finish', weight: 1 },
    { source: 'b', destination: 'a', weight: 3 },
    { source: 'b', destination: 'finish', weight: 5 }
  ]
);

const { distance, previous } = dijkstra(graph, 'start');

console.log(distance);
console.log(previous);

File: original-python-solutions/dijkstra.py
# 1. Найти узел с наименьшей стоимостью
# 2. Проверить, существует ли более дешёвый путь к соседям этого
#    узла, и если существует, обновить их стоимость
# 3. Повторять, пока это не будет сделано для всех узлов графа
# 4. Вычислить итоговый путь

graph = {
  "start": {
    "a": 6,
    "b": 2
  },
  "a": {
    "finish": 1
  },
  "b": {
    "a": 3,
    "finish": 5
  },
  "finish": {}
}

infinity = float("inf")

costs = {
  "a": 6,
  "b": 2,
  "finish": infinity
}

parents = {
  "a": "start",
  "b": "start",
  "finish": None
}

processed = []

def find_lowest_cost_node(costs):
  lowest_cost = float("inf")
  lowest_cost_node = None

  for node in costs:
    cost = costs[node]

    if cost < lowest_cost and node not in processed:
      lowest_cost = cost
      lowest_cost_node = node

  return lowest_cost_node

# Поиск узла с наименьшией стоимостью среди необработанных
node = find_lowest_cost_node(costs)

while node is not None:
  cost = costs[node]
  neighbors = graph[node]

  # Цикл по всем соседям текущего узла
  for key in neighbors.keys():
    new_cost = cost + neighbors[key]

    # Если путь к соседу меньше через текущий узел...
    if costs[key] > new_cost:
      # ... нужно обновить стоимость для этого узла
      costs[key] = new_cost
      # Родителем соседа становится текущий узел
      parents[key] = node

  processed.append(node)

  node = find_lowest_cost_node(costs)

print(costs)
print(parents)
